### PR TITLE
Registrar - visibility and disabled fields

### PR DESCRIPTION
**Pull request by Vojtech-Sassmann**
* For some use-cases, users need to be able to specify fields, that
should be hidden or disabled in some cases. Therefore, new
applicationFormItem fields were created - `hidden` and `disabled`. There
is also a new flag called `updatable`. This flag can be used to
determine if some application field can be updated after submission.
However, some item types cannot still be updated, even if this flag is
set to true (e.g. username).
* It is possible to specify 4 situations, when items should be hidden or
disabled - ALWAYS, NEVER, IF_PREFILLED, IF_EMPTY.
* It is also possible to reference a different item in the form, which
is then used to evaluate the conditions.
* The application types FROM_FEDERATION_SHOW and FROM_FEDERATION_HIDDEN
were merged into a FROM_FEDERATION type. In the dbchangelog, there are
commands, that will transform the old types into the new one, with also
setting the `hidden` and `disabled` columns accordingly.

### CODE REVIEW

**Comment by Vojtech-Sassmann** [comment]
We should probably merge this PR after we test these changes in the Registrar WUI.

File: perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java [hunk 1]
```diff
@@ -2627,7 +2657,7 @@ public void updateFormItemData(PerunSession sess, int appId, ApplicationFormItem
 		ApplicationFormItemData existingData = getFormItemDataById(data.getId(), appId);
 		if (existingData == null) throw new RegistrarException("Form item data specified by ID: "+ data.getId() + " not found or doesn't belong to the application "+appId);
 
-		List<Type> notAllowed = Arrays.asList(FROM_FEDERATION_HIDDEN, FROM_FEDERATION_SHOW, USERNAME, PASSWORD, HEADING, HTML_COMMENT, SUBMIT_BUTTON, AUTO_SUBMIT_BUTTON);
+		List<Type> notAllowed = Arrays.asList(FROM_FEDERATION, USERNAME, PASSWORD, HEADING, HTML_COMMENT, SUBMIT_BUTTON, AUTO_SUBMIT_BUTTON);
```
Comment (balcirakpeter) [inline perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java:2660]: This is not correct. Users can change the prefilled value from fed. He cannot change the value only id the field is Hidden/Disabled. Also if it is possible, add some checks to the submitApplication method, so user does not change disabled/hidden value.

**Review by balcirakpeter — CHANGES_REQUESTED** [review]

File: perun-core/src/main/resources/postgresChangelog.txt [hunk 2]
```diff
@@ -6,6 +6,18 @@
 -- Directly under version number should be version commands. They will be executed in the order they are written here.
 -- Comments are prefixed with -- and can be written only between version blocks, that means not in the lines with commands. They have to be at the start of the line.
 
+3.1.76
```
[the remark this answers is not in the document]
Comment (Vojtech-Sassmann) [reply perun-core/src/main/resources/postgresChangelog.txt:9]: Fixed.

File: perun-db/postgres.sql [hunk 3]
```diff
@@ -486,10 +486,28 @@ create table application_form (
...
+	'ALWAYS',
+	'IF_PREFILLED',
+	'IF_EMPTY'
+	);
+
 -- APPLICATION_FORM_ITEMS - items of application form
 create table application_form_items (
 	id integer not null,
 	form_id integer not null,  --identifier of form (application_form.id)
+	hidden app_item_hidden not null default 'NEVER',
```
[the remark this answers is not in the document]
Comment (Vojtech-Sassmann) [reply perun-db/postgres.sql:507]: Fixed.

File: perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java [hunk 4]
```diff
@@ -2354,7 +2383,7 @@ public void updateFormItem(PerunSession sess, ApplicationFormItem item) throws P
 
 				// We do require value from IDP (federation) if attribute is supposed to be pre-filled and item is required and not editable to users
 				if ((itemW.getPrefilledValue() == null || itemW.getPrefilledValue().isEmpty()) && itemW.getFormItem().isRequired() &&
-						(Type.FROM_FEDERATION_HIDDEN.equals(itemW.getFormItem().getType()) || Type.FROM_FEDERATION_SHOW.equals(itemW.getFormItem().getType()))) {
+						(Type.FROM_FEDERATION == itemW.getFormItem().getType())) {
```
[the remark this answers is not in the document]
Comment (Vojtech-Sassmann) [reply perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java]: Fixed.

File: perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java (hunk 1 above)
Comment (Vojtech-Sassmann) [reply perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java]: Fixed.

File: perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java [hunk 5]
```diff
@@ -2352,17 +2387,20 @@ public void updateFormItem(PerunSession sess, ApplicationFormItem item) throws P
...
-						log.error("Couldn't resolve displayName from: {}, parsedNames were: {}", federValues, parsedName);
-					}
+		Map<Integer, ApplicationFormItemWithPrefilledValue> allItemsByIds = itemsWithValues.stream()
+				.collect(toMap(item -> item.getFormItem().getId(), Function.identity()));
 
-					itemsWithMissingData.add(itemW);
+		for (ApplicationFormItemWithPrefilledValue itemW : itemsWithValues) {
+			// We do require value from IDP (federation) if attribute is supposed to be pre-filled and item is required and not editable to users
+			if ((itemW.getPrefilledValue() == null || itemW.getPrefilledValue().isEmpty()) &&
+					itemW.getFormItem().isRequired() && isItemHidden(itemW, allItemsByIds)) {
```
Comment (balcirakpeter) [inline perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java:2399]: What about disabled items?

**Review by balcirakpeter — CHANGES_REQUESTED** [review]
Please change also the commit message.

File: perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java (hunk 5 above)
Comment (Vojtech-Sassmann) [reply perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java]: Good point, I added a check for that as well.

**Review by balcirakpeter — APPROVED** [review]

**Review by Johaney-s — APPROVED** [review]

**Pull request merged**